File: sdks/python/applad/messaging.py
"""Messaging service."""


class Messaging:
    def __init__(self, client):
        self.client = client

    def send_email(self, to: list, subject: str, html: str = ""):
        body = {"to": to, "subject": subject}
        if html:
            body["html"] = html
        return self.client._call("POST", "/messaging/email", body)

    def send_sms(self, to: list, content: str):
        return self.client._call("POST", "/messaging/sms", {
            "to": to,
            "content": content,
        })

    def send_push(self, to: list, title: str, body: str):
        return self.client._call("POST", "/messaging/push", {
            "to": to,
            "title": title,
            "body": body,
        })

    # --- Templates ---

    def create_template(self, name: str, type_: str, subject: str, body: str,
                        variables: list | None = None, template_id: str = "unique()"):
        """Create a reusable message template with {{variable}} placeholders."""
        return self.client._call("POST", "/messaging/templates", {
            "templateId": template_id,
            "name": name,
            "type": type_,
            "subject": subject,
            "body": body,
            "variables": variables or [],
        })

    def list_templates(self):
        return self.client._call("GET", "/messaging/templates")

    def get_template(self, template_id: str):
        return self.client._call("GET", f"/messaging/templates/{template_id}")

    def update_template(self, template_id: str, **kwargs):
        return self.client._call("PUT", f"/messaging/templates/{template_id}", kwargs)

    def delete_template(self, template_id: str):
        return self.client._call("DELETE", f"/messaging/templates/{template_id}")

    def send_template(self, template_id: str, to: list,
                      variables: dict | None = None):
        """Render the template with variables and send to the given recipients."""
        return self.client._call(
            "POST",
            f"/messaging/templates/{template_id}/send",
            {"to": to, "variables": variables or {}},
        )

File: sdks/dart/lib/src/server/messaging.dart
import '../server_client.dart';

/// Messaging service — send emails, SMS, and push notifications (server-side).
class Messaging {
  final ApplAdServer _client;

  Messaging(this._client);

  Future<Map<String, dynamic>> sendEmail({
    required List<String> to,
    required String subject,
    String? html,
  }) async {
    return _client.post('/v1/messaging/email', data: {
      'to': to,
      'subject': subject,
      if (html != null) 'html': html,
    });
  }

  Future<Map<String, dynamic>> sendSMS({
    required List<String> to,
    required String body,
  }) async {
    return _client.post('/v1/messaging/sms', data: {
      'to': to,
      'body': body,
    });
  }

  Future<Map<String, dynamic>> sendPush({
    required List<String> to,
    required String title,
    required String body,
    Map<String, dynamic>? data,
  }) async {
    return _client.post('/v1/messaging/push', data: {
      'to': to,
      'title': title,
      'body': body,
      if (data != null) 'data': data,
    });
  }

  // --- Templates ---

  Future<Map<String, dynamic>> createTemplate({
    required String name,
    required String type,
    required String subject,
    required String body,
    List<String>? variables,
    String templateId = 'unique()',
  }) {
    return _client.post('/v1/messaging/templates', data: {
      'templateId': templateId,
      'name': name,
      'type': type,
      'subject': subject,
      'body': body,
      'variables': variables ?? [],
    });
  }

  Future<Map<String, dynamic>> listTemplates() {
    return _client.get('/v1/messaging/templates');
  }

  Future<Map<String, dynamic>> getTemplate(String templateId) {
    return _client.get('/v1/messaging/templates/$templateId');
  }

  Future<Map<String, dynamic>> deleteTemplate(String templateId) {
    return _client.delete('/v1/messaging/templates/$templateId');
  }

  Future<Map<String, dynamic>> sendTemplate({
    required String templateId,
    required List<String> to,
    Map<String, String>? variables,
  }) {
    return _client.post('/v1/messaging/templates/$templateId/send', data: {
      'to': to,
      'variables': variables ?? {},
    });
  }
}
